### ISSUE TITLE
nn.init.orthogonal_ doesn't work with multiprocessing

### ISSUE THREAD

**Issue by umichyiwan** [pytorch/pytorch#21956]
## To Reproduce

```
import torch.nn as nn
import torch.multiprocessing as mp
import time
def test():
    print("test in")
    layer = nn.Linear(5 * 5 * 64, 64)
    nn.init.orthogonal_(layer.weight.data)
    print("test out")
processes = []
test()
for rank in range(0, 1):
    p = mp.Process(target=test)
    p.start()
    processes.append(p)
    time.sleep(0.1)
for p in processes:
    time.sleep(0.1)
    p.join()
print('t')
```
output:
test in
test out
test in
t

**Comment by vishwakftw** [comment]
Could you please copy-paste the output from the environment collection script found [here](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)?

**Comment by umichyiwan** [comment]
Collecting environment information...
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.14
GCC version: Could not collect
CMake version: version 3.3.2

Python version: 3.5
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip3] numpy==1.14.5
[pip3] torch==0.4.1
[pip3] torchvision==0.2.1
[conda] torch                     0.4.1                     <pip>
[conda] torchvision               0.2.1                     <pip>

**Comment by vishwakftw** [comment]
Seems to be related: https://github.com/numpy/numpy/issues/654

**Comment by umichyiwan** [comment]
Does pytorch internally use openblas?

**Comment by vishwakftw** [comment]
PyTorch internally uses a flavor of BLAS: MKL, OpenBLAS and so on. The `orthogonal_` initialization uses the QR decomposition internally (which in turn is implemented using `geqrf`, `orgqr` which are LAPACK routines), which is why this issue has popped up.

If you have NumPy installed, could you copy paste the output of `numpy.__config__.show()`?

I ran the script that you have provided and seems to work fine for me. I use a Linux machine with Ubuntu 18.10 w/ CUDA 9.2.

**Comment by umichyiwan** [comment]
Thanks for your response, I am using macbook.

Here is the output:

```
lapack_mkl_info:
  NOT AVAILABLE
openblas_info:
  NOT AVAILABLE
lapack_opt_info:
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
    extra_compile_args = ['-msse3']
atlas_3_10_threads_info:
  NOT AVAILABLE
atlas_3_10_info:
  NOT AVAILABLE
blas_opt_info:
    extra_link_args = ['-Wl,-framework', '-Wl,Accelerate']
    define_macros = [('NO_ATLAS_INFO', 3), ('HAVE_CBLAS', None)]
    extra_compile_args = ['-msse3', '-I/System/Library/Frameworks/vecLib.framework/Headers']
blis_info:
  NOT AVAILABLE
atlas_blas_info:
  NOT AVAILABLE
openblas_clapack_info:
  NOT AVAILABLE
atlas_threads_info:
  NOT AVAILABLE
atlas_3_10_blas_threads_info:
  NOT AVAILABLE
openblas_lapack_info:
  NOT AVAILABLE
atlas_3_10_blas_info:
  NOT AVAILABLE
atlas_blas_threads_info:
  NOT AVAILABLE
atlas_info:
  NOT AVAILABLE
blas_mkl_info:
  NOT AVAILABLE
```

**Comment by vishwakftw** [comment]
Outside of conda, can you try setting OPENBLAS_NUM_THREADS=1 and see if the problem still persists?

**Comment by umichyiwan** [comment]
still persists...

**Comment by nmathew** [comment]
Try this

import torch.nn as nn
import torch.multiprocessing as mp
import time
def test():
    print("test in")
    layer = nn.Linear(5 * 5 * 64, 64)
    nn.init.orthogonal_(layer.weight.data)
    print("test out")
processes = []

if __name__ == '__main__':
    test()
    for rank in range(0, 1):
        p = mp.Process(target=test)
        p.start()
        processes.append(p)
        time.sleep(0.1)
    for p in processes:
        time.sleep(0.1)
        p.join()
    print('t')

**Comment by umichyiwan** [comment]
Thanks, but the problem still exists.

**Comment by umichyiwan** [comment]
Actually if I reduce the size of args of nn.Linear, the problem goes away.

**Comment by ezyang** [comment]
Unfortunately, we're going to need some sort of reproducer if we're going to be able to fix this.

If anyone else is having this problem, please shout!

**Comment by obilaniu** [comment]
@ezyang I have seen this problem before. It's not just `nn.init.orthogonal_()`; Even `torch.zeros()` can provoke this for a "sufficiently large" created Tensor.

## Root Case: `libgomp`

I root-caused it to OpenMP, specifically `libgomp`. When PyTorch or the BLAS it links against require `libgomp`, and the `libgomp` runtime is **first** initialized _(such as by a "sufficiently large" or "sufficiently complex" Tensor operation, which QR is)_ and **then** `fork()`ed without an immediate `exec()`, `libgomp` chokes.

## Diagnosis

The characteristic signs of this condition are:
- `libgomp` is linked (can `grep -e libgomp /proc/<PID>/maps` on the hung process `<PID>`)
- The environment variable `OMP_NUM_THREADS=1` solves the problem
- Hang happens in forked processes

## Solutions

Intel MKL's implementation of OpenMP, `libiomp5`, does not suffer from this problem.
OpenBLAS with the pthreads backend, `libopenblas_pthreads.so` does not suffer from this problem.

Both have in common that they are fork-safe, in that they use the `pthread_atfork()` hack to "properly" shut down their worker threads before forking. 

For this reason, it has been my practice, when building PyTorch from source, to:
- Configure with MKL, if available, otherwise
- Configure with OpenBLAS _with the pthreads backend_ (`libopenblas_pthreads.so`, **not** `libopenblas_openmp.so`), and specify `WITH_OPENMP=0`.

## References

[[1]](https://patchwork.ozlabs.org/patch/319827/), [[2]](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60035), [[3]](https://github.com/xianyi/OpenBLAS/pull/343/files#diff-31f8d4e8863583d95bf2f9529f83844eR258-R274).

**Comment by ezyang** [comment]
@obilaniu Thank you for a really detailed description of the problem. Would you consider this high priority to fix? I can think of two ways we can fix it given your root cause: (1) poison use of omp if we detect we're using a version of a library that doesn't orderly shut down on fork, similar to how we do this for the cuda runtime, (2) make sure our binary packaging doesn't use libgomp (might be easier said than done?)

**Comment by ezyang** [comment]
cc @malfet

**Comment by obilaniu** [comment]
@ezyang The problem is difficult to solve.

- **Some** OpenMP implementations **do** work (Intel MKL's in particular)
- **Some** BLAS implementations **do** work depending on configuration (OpenBLAS)
- **Before** PyTorch is `import`ed, another package could have pulled in `libgomp.so` or another unsafe library and initialized its runtime
- **After** PyTorch is `import`ed, another package could have pulled in `libgomp.so` or another unsafe library and initialized its runtime
- The above two cases can happen **even if** PyTorch itself studiously avoids pulling in `libgomp.so` _itself_.
- There is **no** control over `fork()` except that which `pthread_atfork()` gives.
- There is **no** way to know if `pthread_atfork()` handlers have been invoked because of a `fork()` with the intent to `exec()` immediately after, which is what `"spawn"` fork mode or any `subprocess` call is about, and is safe.

I don't know how one can handle the general case except being careful. There might be easy cases that can be caught, and it might reduce the number of bug reports about inexplicable deadlocks, but catching all cases appears to require too much information for PyTorch to infer from context.

**Comment by ezyang** [comment]
Yeah, I don't think we need to try for a complete solution, just catch the most common error cases.
